### PR TITLE
[RISCV] Add sched model for XiangShan-NanHu

### PR DESCRIPTION
**Pull request by dtcxzyw**
[XiangShan](https://github.com/OpenXiangShan/XiangShan) is an open-source high-performance RISC-V processor.

This PR adds the schedule model for XiangShan-NanHu, the 2nd Gen core of the XiangShan processor series.
Overview: https://xiangshan-doc.readthedocs.io/zh-cn/latest/integration/overview/

It is based on the patch [D122556](https://reviews.llvm.org/D122556) by @SForeKeeper. The original patch hasn't been updated for a long time and it is out of sync with the current RTL design.

Now ICT-CAS is about to complete the tape-out of NanHu core according to @poemonsense. So I posted this PR to add support for it.

[Move elimination](https://github.com/dtcxzyw/llvm-project/commit/59f6e22bf12f67d799a7777f641853fec76c0aa8#diff-32270ea35a510b9a116a50cf5b922c46c194da7b0e0afe76576b8b2bef06556d) and macro fusions will be supported in subsequent PRs.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfoF.td [hunk 1]
```diff
@@ -302,7 +302,7 @@ def FSW : FPStore_r<0b010, "fsw", FPR32, WriteFST32>;
 } // Predicates = [HasStdExtF]
 
 foreach Ext = FExts in {
-  let SchedRW = [WriteFMA32, ReadFMA32, ReadFMA32, ReadFMA32] in {
+  let SchedRW = [WriteFMA32, ReadFMA32, ReadFMA32, ReadFMA32Addend] in {
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVInstrInfoF.td:305]: Should this change be in a separate commit?

File: llvm/lib/Target/RISCV/RISCVInstrInfoD.td [hunk 2]
```diff
@@ -78,7 +78,7 @@ def FSD : FPStore_r<0b011, "fsd", FPR64, WriteFST64>;
 } // Predicates = [HasStdExtD]
 
 foreach Ext = DExts in {
-  let SchedRW = [WriteFMA64, ReadFMA64, ReadFMA64, ReadFMA64] in {
+  let SchedRW = [WriteFMA64, ReadFMA64, ReadFMA64, ReadFMA64Addend] in {
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVInstrInfoD.td:81]: Should this change be in a separate commit?

File: llvm/lib/Target/RISCV/RISCVSchedSiFive7.td [hunk 3]
```diff
@@ -936,7 +936,9 @@ def : ReadAdvance<ReadFMA16, 0>;
 def : ReadAdvance<ReadFMul32, 0>;
 def : ReadAdvance<ReadFMul64, 0>;
 def : ReadAdvance<ReadFMA32, 0>;
+def : ReadAdvance<ReadFMA32Addend, 0>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVSchedSiFive7.td:939]: Please make FMA16 consistent

File: llvm/lib/Target/RISCV/RISCVInstrInfoF.td (hunk 1 above)
Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/RISCVInstrInfoF.td:305]: Posted as #70241.

**Comment by preames** [comment]
Can you separate out the basic processor definition (using NoSchedModel), and a patch which adds the scheduling model?  We can at least get the processor definition landed while we iterate on the scheduling related pieces.

**Comment by dtcxzyw** [comment]
> Can you separate out the basic processor definition (using NoSchedModel), and a patch which adds the scheduling model? We can at least get the processor definition landed while we iterate on the scheduling related pieces.
> 
> edit: For clarity, I'm requesting that the basic processor definition and test updates be made into its own pull request, and that this pull request be reserved for adding the schedule model on top.

Posted as #70294.

File: clang/test/Driver/riscv-cpus.c [hunk 4]
```diff
@@ -20,6 +20,17 @@
 // MCPU-SYNTACORE-SCR1-MAX: "-target-feature" "+zicsr" "-target-feature" "+zifencei"
 // MCPU-SYNTACORE-SCR1-MAX: "-target-abi" "ilp32"
 
+// RUN: %clang --target=riscv64 -### -c %s 2>&1 -mcpu=xiangshan-nanhu | FileCheck -check-prefix=MCPU-XIANGSHAN-NANHU %s
+// MCPU-XIANGSHAN-NANHU: "-nostdsysteminc" "-target-cpu" "xiangshan-nanhu"
+// MCPU-XIANGSHAN-NANHU: "-target-feature" "+m" "-target-feature" "+a" "-target-feature" "+f" "-target-feature" "+d"
```
Comment (MaskRay) [inline clang/test/Driver/riscv-cpus.c:25]: Prefer `-SAME` for strings on the same line

Comment (dtcxzyw) [reply clang/test/Driver/riscv-cpus.c:25]: Fixed. Please move to #70294 to review the processor definition part.

**Comment by dtcxzyw** [comment]
Rebased on top of #70241.

File: llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+def XiangShanNanHuModel : SchedMachineModel {
+  let MicroOpBufferSize = 256;
+  let LoopMicroOpBufferSize = 48;  // Instruction queue size
+  let IssueWidth = 6;  // 6-way decode and dispatch
+  let LoadLatency = 4;
+  let MispredictPenalty = 11; // Based on estimate of pipeline depth.
+  let CompleteModel = 0;
+  let PostRAScheduler = 1; // Enable Post RegAlloc Scheduler pass.
+  let UnsupportedFeatures = [];
```
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:27]: Add `UnsupportedFeatures`? I don't think XiangShan has implement all extensions.

**Review by preames — APPROVED** [review]
LGTM

Note that I'm not reviewing the correctness of the schedule model at all, I'm purely glancing at code structure and testing to make sure this doesn't impact the backend in general.

File: llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+// Zbb
+def : WriteRes<WriteCLZ, [XS2MDU]>;
+def : WriteRes<WriteCLZ32, [XS2MDU]>;
+def : WriteRes<WriteCTZ, [XS2MDU]>;
+def : WriteRes<WriteCTZ32, [XS2MDU]>;
+def : WriteRes<WriteCPOP, [XS2MDU]>;
+def : WriteRes<WriteCPOP32, [XS2MDU]>;
+
+// Zbs
+def : WriteRes<WriteSingleBit, [XS2MDU]>;
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:106]: Is SingleBit and BEXT really 3 cycles? That seems high and could mean it doesn't make sense to use these instructions other than to improve code size. For example BSET is equivalent to LI+SLL+OR which is already 3 cycles.

Comment (poemonsense) [reply llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:106]: I believe only `CLZ(W), CTZ(W), CPOP(W), XPERM(8/4), CLMUL(H/R)` instructions require three cycles

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:106]: > I believe only `CLZ(W), CTZ(W), CPOP(W), XPERM(8/4), CLMUL(H/R)` instructions require three cycles

I believe that these instructions can be implemented to have 1 cycle latency. Is this possible in next-genetion XiangShan-KunMingHu?

Comment (poemonsense) [reply llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:106]: Yes I agree their logics are simple. I think the main reason we were implementing them using higher latencies is the ALU resources (v.s. the use frequency of these instructions). The current ALU design incorporates too many operations, which hurts timing. KMH could re-consider its detailed design to handle these less used instructions. I'm not sure on the schedule of KMH, but possibly @huxuan0307 can see whether KMH can optimize these instructions in the future

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:106]: Thanks! Since RISCV has no dedicated CRC instruction, `Zbc` is very important to accelarate CRC now.

File: llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+
+def XiangShanNanHuModel : SchedMachineModel {
+  let MicroOpBufferSize = 256;
+  let LoopMicroOpBufferSize = 48;  // Instruction queue size
+  let IssueWidth = 6;  // 6-way decode and dispatch
+  let LoadLatency = 4;
+  let MispredictPenalty = 11; // Based on estimate of pipeline depth.
+  let CompleteModel = 0;
+  let PostRAScheduler = 1; // Enable Post RegAlloc Scheduler pass.
+  let UnsupportedFeatures = [HasStdExtZcmt, HasStdExtZkr, HasVInstructions,
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:27]: Hmm, it seems like we need to add a bunch of vendor extensions to the list :(
BTW, `tblgen` complains `No schedule information for instruction 'SM4KS'` due to lack of `UnsupportedSchedZksed`.

[the remark this answers is not in the document]
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:27]: UnsupportedFeatures really doesn't scale very well. Every time a new extensions is added the schedulers need to be updated just to say it isn't supported.

**Comment by dtcxzyw** [comment]
As our internal benchmark shows performance improvement with this PR, I will merge this PR if no more comments.

Happy Chinese New Year!

File: llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td [hunk 8]
```diff
@@ -0,0 +1,308 @@
+//==- RISCVSchedXiangShanNanHu.td - XiangShan-NanHu Scheduling Definitions --*- tablegen -*-=//
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td:1]: Does this line exceed 80 characters?

Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/RISCVSchedXiangShanNanHu.td]: Done.
